javalib/src/main/scala/java/util/StringTokenizer.scala

```diff
@@ -1,103 +1,93 @@
+    var inToken: Boolean = false
+    var i: Int = position
+
+    while(i < length) {
+      if (isDelim(str.charAt(i))) {
+        if (returnDelims)
           count += 1
+
         if (inToken) {
           count += 1
           inToken = false
         }
       } else {
         inToken = true
       }
+
+      i += 1
     }
+
     if (inToken)
       count += 1
+
     count
   }
 
   def hasMoreElements(): Boolean = hasMoreTokens()
 
   def hasMoreTokens(): Boolean = {
-    if (delimiters == null) {
-      throw new NullPointerException()
-    }
-    val length = string.length
-    if (position < length) {
-      if (returnDelimiters)
-        return true
-
-      for (i <- position until length) {
-        if (delimiters.indexOf(string.charAt(i), 0) == -1)
-          return true
+    if (position >= length) false
+    else if (returnDelims) true
+    else {
+      var hasDelim: Boolean = false
+      var i: Int = position
+      while (i < length && !hasDelim) {
+        if (isDelim(str.charAt(i))) hasDelim = true
+        i += 1
       }
+
+      hasDelim
     }
-    false
   }
 
   def nextElement(): Object = nextToken()
 
   def nextToken(): String = {
-    if (delimiters == null) {
-      throw new NullPointerException()
-    }
-    var i      = position
-    val length = string.length
-
-    if (i < length) {
-      if (returnDelimiters) {
-        if (delimiters.indexOf(string.charAt(i), 0) >= 0) {
-          return String.valueOf(string.charAt({ position += 1; position - 1 }))
-        }
-        position += 1
-        while (position < length) {
-          if (delimiters.indexOf(string.charAt(position), 0) >= 0)
-            return string.substring(i, position)
-          position += 1
-        }
-        return string.substring(i)
-      }
+    ensureAvailable()
 
-      while (i < length && delimiters.indexOf(string.charAt(i), 0) >= 0) i += 1
+    if (returnDelims && nextIsDelim) {
+      val ret = String.valueOf(currentChar)
+      position += 1
+      ret
+    } else {
+      // Skip consecutive delims
+      while (position < length && nextIsDelim) position += 1
 
-      position = i
+      ensureAvailable()
 
-      if (i < length) {
-        position += 1
-        while (position < length) {
-          if (delimiters.indexOf(string.charAt(position), 0) >= 0)
-            return string.substring(i, position)
-          position += 1
-        }
-        return string.substring(i)
-      }
+      val start: Int = position
+      while (position < length && !nextIsDelim) position += 1
+      str.substring(start, position)
     }
-    throw new NoSuchElementException()
   }
 
-  def nextToken(delims: String): String = {
-    delimiters = delims
+  def nextToken(delim: String): String = {
+    this.delim = delim
     nextToken()
   }
 
+  private def ensureAvailable(): Unit = {
+    if (position >= length)
+      throw new NoSuchElementException()
+  }
+
+  private def nextIsDelim: Boolean = isDelim(currentChar)
```

Comment:
Is there a reason that nextIsDelim is not also @inline. That way, the predicate is inlined all the way down instead of being a method call to a one-liner.  The SN optimizer may be smart enough to do this already.
I did not count the number of places where nextIsDelim is used, so perhaps there is a code size reason why it is not inlined.  Just makes an inquisitive Reader wonder.